File: recursion/leetcode_Word_Search.java
/**
 * Word_Search.
 *
 * Given a 2D board and a word, find if the word exists in the grid.
 * The word can be constructed from letters of sequentially adjacent cell, where
 * "adjacent" cells are those horizontally or vertically neighboring. The same
 * letter cell may not be used more than once.
 * For example,
 * Given board =
 *
 * [
 * ["ABCE"],
 * ["SFCS"],
 * ["ADEE"]
 * ]
 * word = "ABCCED", -> returns true,
 * word = "SEE", -> returns true,
 * word = "ABCB", -> returns false.
 */


public class leetcode_Word_Search {

  public static void main(String[] args) {
	char[][] board = new char[][] {
      "ABCE".toCharArray(),
      "SFCS".toCharArray(),
      "ADEE".toCharArray()
    };
	System.out.println(exist(board, "ABCCED"));
  }

  public static boolean exist(char[][] board, String word) {
	int n = board.length;
	int m = board[0].length;

	boolean[][] visited = new boolean[n][m];

	for (int i = 0; i < n; ++i) {
	  for (int j = 0; j < m; ++j) {
		if (dfs(i, j, word, visited, board, n, m)) {
		  return true;
		}
	  }
	}

	return false;
  }

  public static boolean dfs(int i, int j, String word, boolean[][] visited, char[][] board, int n, int m) {
	if (word.length() == 0) {
	  return true;
	} else if (i < 0 || i >= n || j < 0 || j >= m || visited[i][j] || word.charAt(0) != board[i][j]) {
	  return false;
	} else {
	  visited[i][j] = true;
	  if (dfs(i + 1, j, word.substring(1), visited, board, n, m)) {
		return true;
	  }
	  if (dfs(i - 1, j, word.substring(1), visited, board, n, m)) {
		return true;
	  }
	  if (dfs(i, j + 1, word.substring(1), visited, board, n, m)) {
		return true;
	  }
	  if (dfs(i, j - 1, word.substring(1), visited, board, n, m)) {
		return true;
	  }
	  visited[i][j] = false;
	  return false;
	}
  }
}

/* Python Version
FYI leetcode OJ for python, the input format is totally fucked up.

def exist(self, board, word):
    def dfs(i, j, word):
        if not word:
            return True
        elif i >= n or i < 0 or j >= m or j < 0 or visited[i][j] or board[i][0][j] != word[0]:
            return False
        else:
            visited[i][j] = True
            if dfs(i + 1, j, word[1:]):
                return True
            if dfs(i - 1, j, word[1:]):
                return True
            if dfs(i, j + 1, word[1:]):
                return True
            if dfs(i, j - 1, word[1:]):
                return True
            visited[i][j] = False
            return False

    n = len(board)
    m = len(board[0][0])
    visited = [ [False for _ in xrange(m) ] for _ in xrange(n) ]

    for i in xrange(n):
        for j in xrange(m):
            if dfs(i, j, word):
                return True

    return False
*/

File: search/leetcode_Word_Search.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode word search
 * Given a 2D board and a word, find if the word exists in the grid.
 *
 * The word can be constructed from letters of sequentially adjacent cell, where
 * "adjacent" cells are those horizontally or vertically neighboring. The same
 * letter cell may not be used more than once.
 *
 */

#include <stdio.h>
#include <string>
#include <utility>
#include <vector>

using std::string;
using std::pair;
using std::vector;


class Solution {
 public:
  bool exist(const vector<vector<char>>& board, string word) {
    if (word.length() == 0) {
      return false;
    }

    vector<vector<bool>> visited (board.size(), vector<bool> (board[0].size(), false));

    for (int i = 0; i < board.size(); ++i) {
      for (int j = 0; j < board[i].size(); ++j) {
        // Start recursion only when:
        // 1. current char is same as first char of word, and
        // 2. length of the word is longer than 1.
        if (board[i][j] == word[0] && (word.length() == 1 || dfs(&visited, pair<int, int> (i, j), board, 0, word))) {
          return true;
        }
      }
    }

    return false;
  }

 private:
  bool dfs(vector<vector<bool>>* visited, pair<int, int> p,
    const vector<vector<char>>& board, int i, string word) {
    if (p.first < 0 || p.first >= visited->size() ||
        p.second < 0 || p.second >= (*visited)[0].size()) {
      // Check point is out of boundary.
      return false;
    } else if (i == word.length()) {
      return true;
    } else if (!(*visited)[p.first][p.second] && word[i] == board[p.first][p.second]) {
      (*visited)[p.first][p.second] = true;
      if (dfs(visited, pair<int, int>(p.first - 1, p.second), board, i + 1, word) ||
          dfs(visited, pair<int, int>(p.first + 1, p.second), board, i + 1, word) ||
          dfs(visited, pair<int, int>(p.first, p.second - 1), board, i + 1, word) ||
          dfs(visited, pair<int, int>(p.first, p.second + 1), board, i + 1, word)) {
        return true;
      }
      // Reset visited bool matrix.
      (*visited)[p.first][p.second] = false;
      return false;
    } else {
      return false;
    }
  }
};


int main() {
  Solution sol;

  vector<vector<char>> board {
    {'A', 'B', 'C', 'E'},
    {'S', 'F', 'C', 'S'},
    {'A', 'D', 'E', 'E'},
  };

  bool exist = sol.exist(board, "ABCCED");
  printf("Exist: %c\n", exist ? 't' : 'f');
}
